### ISSUE TITLE
SC2154 does not recognize variables declared in trap script

### ISSUE THREAD

**Issue by olivergondza** [koalaman/shellcheck#3287]
#### For bugs with existing features

- Rule Id: SC2154
- My shellcheck version (`shellcheck --version` or "online"): 0.11.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

Shellcheck reports variable used in `trap` statement is unassigned even if assigned earlier in the trap "script".

#### Here's a snippet or screenshot that shows the problem:

Mind the trap script is using single quotes - hence the variables are substituted during trap execution, not its declaration.

```bash
#!/usr/bin/env bash

trap 'exit_code=$?; echo >&2 "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $exit_code' ERR
```

```bash
#!/usr/bin/env bash

trap 'let exit_code=$?; echo >&2 "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $exit_code' ERR
```

```bash
#!/usr/bin/env bash

trap 'export exit_code=$?; echo >&2 "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $exit_code' ERR
```

#### Here's what shellcheck currently says:

```
trap 'export exit_code=$?; echo >&2 "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $exit_code' ERR
     ^-- SC2154 (warning): exit_code is referenced but not assigned.
```

#### Here's what I wanted or expected to see:

No errors 😎

**Comment by mattalxndr** [comment]
I'm also running into this. Here's a workaround that doesn't involve a comment:

```bash
trapped() {
    readonly s=$?
    printf >&2 "%s: line %d: %s: error trapped\n" "$0" "$LINENO" "$BASH_COMMAND"
    exit "$s"
}
trap trapped ERR
```
